lib/spack/docs/containers.rst

```diff
@@ -0,0 +1,281 @@
+.. Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+   Spack Project Developers. See the top-level COPYRIGHT file for details.
+
+   SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+.. _containers:
+
+================
+Container Images
+================
+
+Spack can be an ideal tool to setup images for containers since all the
+features discussed in :ref:`environments` can greatly help to manage
+the installation of software during the image build process. Nonetheless,
+building a production image from scratch still requires a lot of
+boilerplate to:
+
+- Get Spack working within the image, possibly running as root
+- Minimize the physical size of the software installed
+- Properly update the system software in the base image
+
+To facilitate users with these tedious tasks, Spack provides a command
+to automatically generate recipes for container images based on
+Environments:
+
+.. code-block:: console
+
+   $ ls
+   spack.yaml
+
+   $ spack containerize
+   # Build stage with Spack pre-installed and ready to be used
+   FROM spack/centos7:latest as builder
+
+   # What we want to install and how we want to install it
+   # is specified in a manifest file (spack.yaml)
+   RUN mkdir /opt/spack-environment \
+   &&  (echo "spack:" \
+   &&   echo "  specs:" \
+   &&   echo "  - gromacs+mpi" \
+   &&   echo "  - mpich" \
+   &&   echo "  concretization: together" \
+   &&   echo "  config:" \
+   &&   echo "    install_tree: /opt/software" \
+   &&   echo "  view: /opt/view") > /opt/spack-environment/spack.yaml
+
+   # Install the software, remove unecessary deps
+   RUN cd /opt/spack-environment && spack install && spack gc -y
+
+   # Strip all the binaries
+   RUN find -L /opt/view/* -type f -exec readlink -f '{}' \; | \
+       xargs file -i | \
+       grep 'charset=binary' | \
+       grep 'x-executable\|x-archive\|x-sharedlib' | \
+       awk -F: '{print $1}' | xargs strip -s
+
+   # Modifications to the environment that are necessary to run
+   RUN cd /opt/spack-environment && \
+       spack env activate --sh -d . >> /etc/profile.d/z10_spack_environment.sh
+
+
+   # Bare OS image to run the installed executables
+   FROM centos:7
+
+   COPY --from=builder /opt/spack-environment /opt/spack-environment
+   COPY --from=builder /opt/software /opt/software
+   COPY --from=builder /opt/view /opt/view
+   COPY --from=builder /etc/profile.d/z10_spack_environment.sh /etc/profile.d/z10_spack_environment.sh
```

Comment:
Could there be an option to copy spack too, so the final image can have a working spack environment?

Reply:
What would be the use case for that?

Reply:
The use case I have in mind is support for base images. Later layers may still need a way to install packages.

Reply:
I don't think it's a good idea to copy Spack over, since that basically means having to recreate the build stage (so why bother with a multi-stage build in the first place?). Also copying Spack is not the full story as there are a bunch of system packages that needs to be installed too and increase the size of the image.

If the use case use case is motivated by installing more software into an existing image I would modify the corresponding `spack.yaml` and rebuild the image entirely. Also, in case the build stage is really needed it can be tagged:
```docker
$ docker build --target build -t <sometag>
```
Does that cover your needs?

Reply:
> That probably doesn't align with what this PR is trying to accomplish.

Couldn't the site just provide an initial `spack.yaml` (with MPI, etc. already in there) that the user can then modify by e.g. adding other libraries or applications?

Reply:
Updates to that base yaml would require a downstream to hand re-write theirs, which is error prone and would make automatic build pipelines more difficult. It's far easier to target a base image and change the version number, or to simply use the latest tagged release. Moreover, configuration that goes into a base image can be obscured from a downstream dev until they actively look for it.

Reply:
@alalazo: I actually understand and think @sknigh has a really good use case here, but maybe we should hold off on those modifications for another PR.

There is some dissonance between what we're trying to do here and with how our Spack base images are created.  It would be cool if we could unify this command with the various scripts we have to create base images, so that people could create either a good base image or a stripped-down app image.  I think sorting that out is a separate topic though -- this is a first cut and the main ECP deliverable we've got is small, multi-stage builds.

@sknigh: want to open an issue with a description of the workflows you want to support, and proposed modifications (high level) to the command?